File: seen.pl
use strict;

my ($who, $nick) = @ARGV;
my $DAT_FILE = "seen.dat";

$| = 1;

unless (open SEEN, "<$DAT_FILE") {
	print STDERR "$DAT_FILE: $!\n";
	print "Pole siin kedagi.\n";
	exit 1
}

our %seen;

while (<SEEN>) {
	my @r = split /\t/s;
	unless ($#r) {
		$_->[1] =~ s/^\+//s for values %seen;
		next
	}
	our $k = lc $r[0];
	if (length $r[2] <= 1 and exists $seen{$k}) {
		$_ = $seen{$k}->[2];
		$r[2] = $_ if $_;
	}
	$seen{$k} = \@r;
}

# write compacted seen.dat
# this is because the main bot just appends to the seen.dat
open TMP, ">$DAT_FILE.tmp" or die "Cannot create temp";
print TMP join("\t", @{$_}) for values %seen;
close TMP;

rename "$DAT_FILE.tmp", $DAT_FILE or print STDERR "$DAT_FILE: $!\n";
close SEEN;

if (lc $who eq lc $nick) {
	print "$nick: Paraku oled sa ikka veel siin";
	exit
}

my $r = $seen{lc $nick};
if ($r) {
	$r->[1] =~ /^(\+?)(.*)/s;
	my $alive = $1;
	if ($alive) {
		print "$r->[0] on praegu kanalil"
	} else {
		my $t = time - $2;
		my $s = $t % 60;
		my $t = int($t / 60);
		my $m = $t % 60;
		my $t = int($t / 60);
		my $h = $t % 24;
		my $t = int($t / 24);
		print ("$r->[0] oli siin ${t}d ${h}h ${m}m ${s}s tagasi" .
				($r->[2] =~ /\S/ ? " ja ütles: $r->[2]" : "\n"))
	}
} else {
	print "Ei tea midagi ${nick}'st\n"
}

File: seen.nim
import std/[strutils, os, posix, tables, times]

let who = paramStr(1)
let nick = paramStr(2)
let datFile = if paramCount() < 3: "seen.dat"
              else: paramStr(3)

let seenFile = try: datFile.open
               except IOError as e:
                echo "Pole siin kedagi."
                quit e.msg

var seen = initTable[string, seq[string]]()
var count = 0
var line: string

# each line is "nick\tTIME\tMESSAGE", online nicks have +TIME
while seenFile.readLine(line):
    inc count
    var r = line.split('\t')
    if r.len <= 1: # Connected IRC, therefore all previous seens are offline
        for v in seen.mvalues:
            v[1].removePrefix '+'
    else:
        let key = r[0].toLowerAscii
        if r[2].len == 0 and key in seen: # steels message
            r[2] = seen[key][2]
        seen[key] = r
seenFile.close

# write compacted seen.dat
# this is because the main bot just appends to the seen.dat
if seen.len < count:
    let tmpName = datFile & ".tmp"
    let tmpFile = tmpName.open fmAppend # fmAppend to not truncate before locking
    var lock = Tflock(l_type: static[cshort](F_WRLCK))
    let fd = tmpFile.getOsFileHandle
    if fcntl(fd, F_SETLK, addr lock) != -1:
        # now that we have lock, ensure that the file is empty
        discard ftruncate(fd, 0)
        for entry in seen.values:
            tmpFile.writeLine entry.join("\t")
        tmpFile.flushFile
        if fdatasync(fd) == 0:
            moveFile tmpName, datFile
        else:
            discard tryRemoveFile tmpName
        tmpFile.close
    else:
        tmpFile.close

let lowerNick = nick.toLowerAscii
if who.toLowerAscii == lowerNick:
    echo nick & ": Paraku oled sa ikka veel siin"
    quit 0

let r = seen.getOrDefault lowerNick
if r == []:
    echo "Ei tea midagi " & nick & "'st"
elif r[1].startsWith '+':
    echo r[0] & " on praegu kanalil"
else:
    let diff = toInt(epochTime() - r[1].parseFloat)
    var t = abs diff
    let s = t mod 60
    t = t div 60
    let m = t mod 60
    t = t div 60
    let h = t mod 24
    t = t div 24 * 400
    let d = t mod 146097 div 400
    let y = t div 146097
    var msg = r[0] & " oli siin "
    if y != 0: msg &= $y & "a "
    if d != 0: msg &= $d & "p "
    if h != 0: msg &= $h & "t "
    if m != 0: msg &= $m & "m "
    if y == 0: msg &= $s & "s "
    if diff >= 0: msg &= "tagasi"
    else: msg &= "pärast"
    if r[2] != "": msg &= " ja ütles: " & r[2]
    echo msg
